subsys/net/lib/lwm2m/lwm2m_rd_client.c

```diff
@@ -0,0 +1,874 @@
+				break;
+
+			case ENGINE_DEREGISTER_FAILED:
+				break;
+
+			case ENGINE_DEREGISTERED:
+				break;
+
+			default:
+				SYS_LOG_ERR("Unhandled state: %d",
+					    get_sm_state(index));
+
+			}
+
+			k_yield();
+		}
+
+		/*
+		 * TODO: calculate the diff between the start of the loop
+		 * and subtract that from the update interval
+		 */
+		k_sleep(K_MSEC(STATE_MACHINE_UPDATE_INTERVAL));
+	}
+}
+
+static bool peer_addr_exist(struct sockaddr *peer_addr)
+{
+	bool ret = false;
+	int i;
+
+	/* look for duplicate peer_addr */
+	for (i = 0; i < client_count; i++) {
+#if defined(CONFIG_NET_IPV6)
+		if (peer_addr->family == AF_INET6 && net_ipv6_addr_cmp(
+				&net_sin6(&clients[i].bs_server)->sin6_addr,
+				&net_sin6(peer_addr)->sin6_addr)) {
+			ret = true;
+			break;
+		}
+
+		if (peer_addr->family == AF_INET6 && net_ipv6_addr_cmp(
+				&net_sin6(&clients[i].reg_server)->sin6_addr,
+				&net_sin6(peer_addr)->sin6_addr)) {
+			ret = true;
+			break;
+		}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+		if (peer_addr->family == AF_INET && net_ipv4_addr_cmp(
+				&net_sin(&clients[i].bs_server)->sin_addr,
+				&net_sin(peer_addr)->sin_addr)) {
+			ret = true;
+			break;
+		}
+
+		if (peer_addr->family == AF_INET && net_ipv4_addr_cmp(
+				&net_sin(&clients[i].reg_server)->sin_addr,
+				&net_sin(peer_addr)->sin_addr)) {
+			ret = true;
+			break;
+		}
+#endif
+	}
+
+	return ret;
+}
+
+static void set_ep_ports(int index)
+{
+#if defined(CONFIG_NET_IPV6)
+	if (clients[index].bs_server.family == AF_INET6) {
+		net_sin6(&clients[index].bs_server)->sin6_port =
+			htons(LWM2M_BOOTSTRAP_PORT);
+	}
+
+	if (clients[index].reg_server.family == AF_INET6) {
+		net_sin6(&clients[index].reg_server)->sin6_port =
+			htons(LWM2M_PEER_PORT);
+	}
+#endif
+
+#if defined(CONFIG_NET_IPV4)
+	if (clients[index].bs_server.family == AF_INET) {
+		net_sin(&clients[index].bs_server)->sin_port =
+			htons(LWM2M_BOOTSTRAP_PORT);
+	}
+
+	if (clients[index].reg_server.family == AF_INET) {
+		net_sin(&clients[index].reg_server)->sin_port =
+			htons(LWM2M_PEER_PORT);
+	}
+#endif
+}
+
+int lwm2m_rd_client_start(struct net_context *net_ctx,
+			  struct sockaddr *peer_addr,
+			  const char *ep_name)
+{
+	int index;
+
+	if (client_count + 1 > CLIENT_INSTANCE_COUNT) {
+		return -ENOMEM;
+	}
+
+	if (peer_addr_exist(peer_addr)) {
+		return -EEXIST;
+	}
+
+	/* Server Peer IP information */
+	/* TODO: use server URI data from security */
+	index = client_count;
+	client_count++;
+	clients[index].net_ctx = net_ctx;
+	memcpy(&clients[index].reg_server, peer_addr, sizeof(struct sockaddr));
+	memcpy(&clients[index].bs_server, peer_addr, sizeof(struct sockaddr));
+	set_ep_ports(index);
+	set_sm_state(index, ENGINE_INIT);
+	strncpy(clients[index].ep_name, ep_name,
+		sizeof(clients[index].ep_name)-1);
+	SYS_LOG_INF("LWM2M Client: %s", clients[index].ep_name);
+
+	return 0;
+}
+
+static int lwm2m_rd_client_init(struct device *dev)
+{
+	k_thread_create(&lwm2m_rd_client_thread_data,
```

Comment:
There are lot of threads created by lwm2m system, is it possible to combine them (or some of them)? These many threads use quite a lot of precious ram for stack space.